### PR TITLE
GitHub Issue NOAA-EMC/GSI#270. GSI code change to assimilate Saildrone, Argo and Glider in GFS NSST

### PR DESCRIPTION
**Pull request by XuLi-NOAA**
A GSI code change to assimilate three types of in situ sea water temperature observation is ready to merge. See https://github.com/NOAA-EMC/GSI/issues/270

### CODE REVIEW

**Comment by MichaelLueken-NOAA** [comment]
Hi @XuLi-NOAA,

It appears that there might have been an issue while preparing your fork to create the PR for issue #270.  Please use the following commands to correct the PR:

1) git clone --recursive git@github.com:XuLi-NOAA/GSI.git update
2) cd update
3) git reset --hard HEAD~1
4) git reset --hard HEAD~1
5) git reset --hard HEAD~1
6) git reset --hard HEAD~1
7) git remote add upstream https://github.com/NOAA-EMC/GSI.git
8) git remote update
9) git rebase upstream/master
10) git submodule update
11) git checkout feature/GSI_nsstdata
12) git checkout master
13) git merge --squash feature/GSI_nsstdata
14) `git commit -m "GitHub Issue NOAA-EMC/GSI#270. Add capability to assimilate Saildrone, Argo, & Glider observation in GSI: NSST"`
15) git push origin master --force

Once these steps are done, the PR should be ready to go.  Please let me know who you would like to have review these changes.

Thanks,

Mike

**Comment by XuLi-NOAA** [comment]
Hi, Mike,

Done.

And Andrew and Emily can be the reviewers.

Thanks!

Can you tell why this happened? Because we did something special last time?

Xu

**Comment by XuLi-NOAA** [comment]
No review is done so far. It needs hurry if this will be included in 16.x.

File: src/gsi/read_nsstbufr.f90 [hunk 1]
```diff
@@ -260,6 +279,31 @@ subroutine read_nsstbufr(nread,ndata,nodata,gstime,infile,obstype,lunout, &
...
+        elseif ( trim(subset) == 'NC031005' ) then            ! Argo or Glider
+           msst = 12.0_r_kind                                 ! for ARGO, assign to be 12
+           call ufbint(lunin,tpf3,3,65535,klev,'SSTH SALNH WPRES') ! get klev,T,water pres & salinity
+
+           call ufbint(lunin,r_rpid,1,1,iret,'WMOP')
+           write(crpid,'(I7)') int(r_rpid)
+
+           apres(:) = real(tpf3(3,:))
+
+           pres1 = minval(apres)/10000.  ! converting wpres from Pa to dbar for minmum depth
```
Comment (MichaelLueken-NOAA) [inline src/gsi/read_nsstbufr.f90:291]: Please add precision to real value, 10000.

File: src/gsi/read_nsstbufr.f90 [hunk 2]
```diff
@@ -260,6 +279,31 @@ subroutine read_nsstbufr(nread,ndata,nodata,gstime,infile,obstype,lunout, &
...
+
+           apres(:) = real(tpf3(3,:))
+
+           pres1 = minval(apres)/10000.  ! converting wpres from Pa to dbar for minmum depth 
+                                                  ! Not always layer 1 
+!
+!          get zob in meters
+!
+           zob = ((-3.434e-12*pres1+1.113e-7)*pres1+0.712953)*pres1 + 14190.7*log(1.0+1.83e-5*pres1)
+           zob = (zob/(980.+1.113e-4*pres1))*1000.
```
Comment (MichaelLueken-NOAA) [inline src/gsi/read_nsstbufr.f90:297]: Please add precision to real values -3.434e-12, 1.113e-7, 0.712953, 14190.7, 1.0, 1.83e-5, 980., 1.113e-4, and 1000.

**Review by MichaelLueken-NOAA — COMMENTED** [review]
I have completed my visual review of the changes.  There are two issues that will need to be addressed before I can send this work out to the review committee - first, please remove the issue #303 changes from this PR and add proper precision to real values (i.e., adding _r_kind to 10000. values).

To remove issue #303 changes from this PR (and remove the two extra commit messages):

1) git clone --recursive git@github.com:XuLi-NOAA/GSI.git update
2) cd update
3) git reset --hard HEAD~1
4) git reset --hard HEAD~1

Now, your master is back to the correct one update associated with this PR.  To update your master to the latest in the authoritative repository:

1) git remote add upstream https://github.com/NOAA-EMC/GSI.git
2) git remote update
3) git rebase upstream/master

Your fork's master branch is now up-to-date with the authoritative repository. 

From this point, please add the proper precision to the real values noted in my review, then commit using:

1) git add read_nsstbufr.f90
2) git commit --amend
3) Please save and close commit message window without altering commit message
4) git push origin master --force

Once these steps are complete and I have finished the regression and debug testing, I should be able to submit this work to the review committee with a quick turn around (only a single file has been updated).

If you have any questions, please let me know.

**Comment by MichaelLueken-NOAA** [comment]
Hi @XuLi-NOAA, I have completed my regression tests and the global_fv3_4denvar_C192 test failed due to reproducibility.  I'm checking to see if this is expected.  If it is, then there are no issues.  If it isn't, then this reproducibility failure will need to be addressed:

global_fv3_4denvar_C192_regression_results.txt:thus the regression test has Failed on cost for global_fv3_4denvar_C192_loproc_updat and global_fv3_4denvar_C192_loproc_contrl analyses.
global_fv3_4denvar_C192_regression_results.txt:Thus, the case has Failed siganl of the regression tests.

**Comment by MichaelLueken-NOAA** [comment]
@XuLi-NOAA Yes, -3.434e-12 would be -3.434e-12_r_kind

**Comment by XuLi-NOAA** [comment]
@MichaelLueken-NOAA : I have done the two issues you pointed.

File: src/gsi/read_nsstbufr.f90 [hunk 3]
```diff
@@ -260,6 +279,31 @@ subroutine read_nsstbufr(nread,ndata,nodata,gstime,infile,obstype,lunout, &
...
+           write(crpid,'(I7)') int(r_rpid)
+
+           apres(:) = real(tpf3(3,:))
+
+           pres1 = minval(apres)/10000.0_r_kind   ! converting wpres from Pa to dbar for minmum depth 
+                                                  ! Not always layer 1 
+!
+!          get zob in meters
+!
+           zob = ((-3.434e-12_r_kind*pres1+1.113e-7_r_kind)*pres1+0.712953_r_kind)*pres1 + 14190.7_r_kind*log(one+1.83e-5_r_kind*pres1)
```
Comment (MichaelLueken-NOAA) [inline src/gsi/read_nsstbufr.f90:296]: The coding standards call for no more than 132 characters per line.  Line 296 has 135 characters.  Please break up this line to adhere to the 132 character limit.

**Review by MichaelLueken-NOAA — COMMENTED** [review]
Hi Xu,

I was able to clone your update and it built (in both standard and debug) without issue.  I did note that, after adding the _r_kind to the real values on line 296, the 132 character limit per line has been broken. Please break up line 296 so that the line doesn't exceed the 132 character limit.

Mike

File: src/gsi/read_nsstbufr.f90 (hunk 3 above)
Comment (XuLi-NOAA) [reply src/gsi/read_nsstbufr.f90]: @MichaelLueken-NOAA : Done

**Comment by MichaelLueken-NOAA** [comment]
The due date for feedback from the DA review committee has passed, so I will now give final approval to this update and merge it to the authoritative repository.

**Pull request merged**